src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -195,6 +197,15 @@ var ReactCompositeComponentMixin = {
           Component.displayName || Component.name || 'Component'
         );
       }
+
+      propsMutated = !shallowEqual(inst.props, publicProps);
```

Comment:
This should use `===` not shallowEqual for performance (and better semantics, IMO) and this check should be dev-only.